File: common/src/main/java/com/nihility/utils/Singleton.java
package com.nihility.utils;

import androidx.annotation.NonNull;

import java.lang.reflect.Constructor;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class Singleton {
    static Map<Class<?>, Object> instances = new ConcurrentHashMap<>();
    static Map<Class<?>, Object> userInstances = new ConcurrentHashMap<>();

    public static <T> T instance(T... reified) {
        if (reified.length > 0) {
            throw new IllegalArgumentException(
                    "Please don't pass any values here. Java will detect class automagically.");
        }
        Class<T> klass = getClassOf(reified);

        T obj = (T) userInstances.get(klass);
        if (obj != null) {
            return obj;
        }

        obj = (T) instances.get(klass);
        if (obj != null) {
            return obj;
        }

        return create(klass);
    }

    private static synchronized <T> @NonNull T create(Class<T> klass) {
        T obj = (T) instances.get(klass);
        if (obj != null) {
            return obj;
        }
        try {
            Constructor<T> constructor = klass.getDeclaredConstructor();
            constructor.setAccessible(true);
            obj = constructor.newInstance();
            instances.put(klass, obj);
            return obj;
        } catch (Throwable e) {
            throw new RuntimeException(e);
        }
    }

    public static <T> AutoReset reset(T... reified) {
        Class<T> klass = getClassOf(reified);

        if (reified.length == 0) {
            userInstances.remove(klass);
        } else {
            userInstances.put(klass, reified[0]);
        }
        return new AutoReset(klass);
    }

    public static class AutoReset implements AutoCloseable {
        Class<?> klass;
        AutoReset(Class<?> klass) {
            this.klass = klass;
        }

        @Override
        public void close() {
            Singleton.userInstances.remove(klass);
        }
    }

    private static <T> Class<T> getClassOf(T[] array) {
        return (Class<T>) array.getClass().getComponentType();
    }
}

File: common/src/main/java/com/nihility/utils/Singleton.kt
@file:Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")
package com.nihility.utils

import java.util.concurrent.ConcurrentHashMap

object Singleton {
    @PublishedApi
    internal val instances = ConcurrentHashMap<Class<*>, Any>()

    @PublishedApi
    internal val userInstances = ConcurrentHashMap<Class<*>, Any>()

    @Suppress("UNCHECKED_CAST")
    inline fun <reified T : Any> instance(): T {
        val klass = T::class.java

        val userObj = userInstances[klass] as? T
        if (userObj != null) {
            return userObj
        }

        val obj = instances[klass] as? T
        if (obj != null) {
            return obj
        }

        return create(klass)
    }

    @JvmStatic
    @Suppress("UNCHECKED_CAST")
    fun <T : Any> instance(vararg reified: T): T {
        if (reified.isNotEmpty()) {
            throw IllegalArgumentException(
                "Please don't pass any values here. Java will detect class automagically."
            )
        }
        val klass = reified.javaClass.componentType as Class<T>
        return get(klass)
    }

    @JvmStatic
    @Suppress("UNCHECKED_CAST")
    fun <T : Any> get(klass: Class<T>): T {
        val userObj = userInstances[klass] as? T
        if (userObj != null) {
            return userObj
        }

        val obj = instances[klass] as? T
        if (obj != null) {
            return obj
        }

        return create(klass)
    }

    @JvmStatic
    @Synchronized
    @Suppress("UNCHECKED_CAST")
    @PublishedApi
    internal fun <T : Any> create(klass: Class<T>): T {
        var obj = instances[klass] as? T
        if (obj != null) {
            return obj
        }
        return try {
            val constructor = klass.getDeclaredConstructor()
            constructor.isAccessible = true
            obj = constructor.newInstance()
            instances[klass] = obj as Any
            obj
        } catch (e: Throwable) {
            throw RuntimeException(e)
        }
    }

    @Suppress("UNCHECKED_CAST")
    inline fun <reified T : Any> reset(value: T? = null): AutoReset {
        val klass = T::class.java
        if (value == null) {
            userInstances.remove(klass)
        } else {
            userInstances[klass] = value
        }
        return AutoReset(klass)
    }

    @JvmStatic
    fun <T : Any> reset(vararg reified: T): AutoReset {
        val klass = reified.javaClass.componentType as Class<T>
        if (reified.isEmpty()) {
            userInstances.remove(klass)
        } else {
            userInstances[klass] = reified[0] as Any
        }
        return AutoReset(klass)
    }

    @JvmStatic
    fun <T : Any> reset(klass: Class<T>, value: T?): AutoReset {
        if (value == null) {
            userInstances.remove(klass)
        } else {
            userInstances[klass] = value
        }
        return AutoReset(klass)
    }

    class AutoReset(private val klass: Class<*>) : AutoCloseable {
        override fun close() {
            userInstances.remove(klass)
        }
    }
}
